specs/agents/tracing-instrumentation-messaging.md

```diff
@@ -188,6 +190,30 @@ queues/topics/exchanges will be ignored.
 | Central config | `true` |
 
 
+### Message metadata
+
+To support distributed tracing with automatic instrumentation, the messaging
+system must provide a mechanism to add metadata/properties/attributes to
+individual messages, akin to HTTP headers. If an APM agent supports
+trace-context for a given messaging system, it MUST use the following mechanisms
+so that cross-language tracing works:
+
+| Messaging system       | Mechanism |
+| ---------------------- | --------- |
+| Azure Queue            | ??? |
+| Azure Service Bus      | ??? |
+| Java Messaging Service | ??? |
+| Apache Kafka           | TODO: is the Java APM agent using [Kafka Record headers](https://cwiki.apache.org/confluence/display/KAFKA/KIP-82%2B-%2BAdd%2BRecord%2BHeaders)? |
+| RabbitMQ               | TODO: are [Message Attributes](https://www.rabbitmq.com/tutorials/amqp-concepts.html#messages) feasible? |
```

Comment:
Yes. See also https://github.com/elastic/apm-agent-java/blob/main/apm-agent-plugins/apm-rabbitmq/apm-rabbitmq-plugin/src/main/java/co/elastic/apm/agent/rabbitmq/header/RabbitMQTextHeaderGetter.java.

In the Java API, message attributes are referred to as `AMQP.BasicProperties` but it seems to be the same thing.

Reply:
Thanks. Commit 13dc897